### PR TITLE
Check byKey in transaction validation

### PR DESCRIPTION
**Pull request by cocreature**
No tests because validation has no tests in general :(

part of #7622

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala [hunk 1]
```diff
@@ -227,7 +240,7 @@ private final class Validation[Nid, Cid](implicit ECid: Equal[Cid]) {
                     signatories2,
                     stakeholders2,
                     key2,
-                    byKey2 @ _,
+                    byKey2,
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:243]: Good catch. I wondering who write the original code ;).

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM. Thanks.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala [hunk 2]
```diff
@@ -237,7 +250,8 @@ private final class Validation[Nid, Cid](implicit ECid: Equal[Cid]) {
                   (actingParties1.isEmpty || actingParties1 == actingParties2) &&
                   signatories1 == signatories2 &&
                   stakeholders1 == stakeholders2 &&
-                  (key1.isEmpty || keyIsReplayedBy(key1, key2)) =>
+                  (key1.isEmpty || keyIsReplayedBy(key1, key2)) &&
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Validation.scala:253]:
```suggestion
                  (key1. isEmpty || byKeyIsReplayedBy(byKey, byKey2)) &&
```

File: daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala [hunk 3]
```diff
@@ -95,12 +95,23 @@ private[preprocessing] final class TransactionPreprocessor(
...
             case exe: Node.NodeExercises[_, Cid] =>
-              val (cmd, newCids) = commandPreprocessor.unsafePreprocessExercise(
-                exe.templateId,
-                exe.targetCoid,
-                exe.choiceId,
-                exe.chosenValue,
-              )
+              import scala.Ordering.Implicits.infixOrderingOps
+              val (cmd, newCids) = exe.key match {
+                case Some(key) if exe.byKey && exe.version >= TransactionVersion.minByKey =>
```
Comment (cocreature) [inline daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:100]: @remyhaemmerle-da can you please check if seems sensible? This was (understandably) necessary to fix the conformance tests.

Comment (cocreature) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:100]: > I would prefer we do check for version, i.e. the transaction coder ensures already that exe.byKey is defined only if exe.version >= TransactionVersion.minByKey

That seems fragile. Now it matters if I first roundtrip via serialization & deserialization or whether I get a transaction directly out of the engine and pass it here. I don’t really see a usecase for the second case but I still feel slightly uneasy about breaking it.

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:100]: The validation was original designed to work only on deserialized transaction. I think  we should continue to assume this as: 
- it simplify reasoning about validation as transaction are somehow normalized by the serialization/deserialization process. 
- I do not see use case were we may want to validate non serialized transaction
- The way we are distributing the versioning logic through the code seems to me very fragile, I really want to limit it to some very limited places. You introduce in this PR tow new places.  
- why changing something that has proven to work.

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:100]: I would be fine with writing a transaction (version aware) normalize equivalent to the serialization-deserialization.

Comment (remyhaemmerle-da) [reply daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/TransactionPreprocessor.scala:100]: I was indeed thinking to apply it inside the finish. For canton I see several solution:
 - We add a config flag to normalize or not the `byKey` flag
 - We include the `byKey` flag info in the `Transaction.MetaData`
 - We force Canton to use only LF 1.14 (wish could be somehow possible as there is no Canton production).

**Pull request merged**